### ISSUE TITLE
[Architecture] Refactoring OptMeowt

### ISSUE THREAD

**Issue by kalicki1** [privacy-tech-lab/optmeowt-browser-extension#166]
I believe this issue has the highest priority of those mentioned in #165. In short, we want to figure out how we will restructure OptMeowt to make more sense as a larger, more intricate project, as well as one that will continue to grow in the coming months. Given we want to make the codebase easy to understand, as well as add new analysis functionality, we should modularize the existing code with good organization as done in similar privacy analysis extensions. We can accomplish this by using a combination of technologies such as Webpack & Babel (for web compatibility and to help facilitate modularity).

In terms of concrete structure, we discussed the following general hierarchy (this can be updated to reflect what we have as work is done). Note that our original diagram is in issue #165. 

```
.
├── src                 
|   ├── assets          # Other files (photos, media, etc.)
|   ├── background      # Main code
|   |   ├── oldcode     # # Refactored old OptMeowt code
|   |   └── analysis    # # new analysis code — to be developed -
|   ├── json            # JSON files (user preferences)
|   ├── libs            # Libraries
|   ├── options         # Frontend (Preferences page)
|   ├── popup           # Frontend (Popup)
|   └── manifest.json   # metadata
└── ...
```

In addition, as far as we have decided, we will be continuing with a "two mode" approach that consists of our usual OptMeowt functionality, as well as an "analysis" mode that runs when requested to do so. This is the reason for organizing our old code in parallel with the analysis code. 

The above `oldcode` folder, which for now will hold the existing, refactored OptMeowt code, will need to be broken down even further once we get going. 

Please feel free to add suggestions as well! I will open a new branch for the refactored project.

**Comment by SebastianZimmeck** [comment]
Looks good! Can't think of any improvements at this point.

**Comment by kalicki1** [comment]
The last commit is the first step in refactoring our code. Currently the extension will only send headers, however I moved all of the functions for our main background listeners into the `events.js` file for simplicity and readability. Moving forward I will add each component of the extension piece by piece and push the changes as I get them done. To keep track of what I intend to do, I hope to:

- [ ] Create wrapper for storage (to make implementing DB to replace localstorage easier)
- [ ] Change JSON files to consts exported from .js files
- [ ] Add global settings page (for builds for Firefox vs. Chrome)
- [ ] Look into using Babel/Webpack in order to use Node modules & abstract what version we build
- [ ] Similarly, look into task runners (NPM vs grunt) for helping with building Firefox vs. Chrome versions

For now, I moved our info in `headers.json` to `headers.js` and use its exported object in our code.

**Comment by kalicki1** [comment]
For some reason, my current GPC DOM property code doesn't work as expected on Chrome... I get a failed to load dom.js error thrown (while running via chrome.runtime.executeScript). I will work around this for now, and in the future if it is not resolved we can change the behavior and run it off of a content script.

**Comment by kbeliauski** [comment]
@kalicki1 pushed the changes, and I am going to try to implement IndexedBD for the domain list. we are working on the way to use the wrapper function to use everywhere in the extension code

**Comment by kbeliauski** [comment]
Should I create a new branch just for me messing around with the IndexedDB? And I can put updates in the corresponding issue

**Comment by SebastianZimmeck** [comment]
@kbeliauski, yes, if you are making changes to the codebase based on this issue, it should be in the `issue-166` branch (which, I believe, you already opened).

Each related commit message should then reference `issue #166`.

**Comment by kalicki1** [comment]
So far, I have managed to connect the backend local storage back to our extension 🥳 This pertains to issue #178 which I will close. @kbeliauski is still working on implementing indexedDB which should be a very hidden change since I have created wrappers that abstract the storage method in issue #178. So far, the cleaned up code contains 

- `background.js`
- `events.js`
- `dom.js`
- `domainlist.js`
- `storage.js`

I will sketch a diagram of how everything flows to get a sense of the larger architecture in the coming days.

**Comment by SebastianZimmeck** [comment]
> I will sketch a diagram of how everything flows to get a sense of the larger architecture in the coming days.

Excellent! That will be very useful.

**Comment by kalicki1** [comment]
I just pushed a change that allows us to use Webpack in OptMeowt. It is still rough around the edges and will need to be cleaned up, but it should allow work to progress in our new development environment. We now have npm package support!

For the time being, in order to run and test the extension, make sure to have npm installed and then run `npm run build` while pointing to the root directory to tell npm to run our build command which in turn calls Webpack according to our `webpack.config.js` file. Then, open the extension from the `dist` folder. 

@stanleymarkman @kbeliauski feel free to play around with the Webpack config file as well as the npm task runner scripts. There may be better ways of doing different things than I am at the moment.

**Comment by kalicki1** [comment]
I will work on adding support for compiling our html documents in addition to our main js functionality.

**Comment by kalicki1** [comment]
Just pushed an update to move our JSON files containing information about opt-out signals from `src/json` to `src/data` as const exports inside of .js files.

**Comment by kalicki1** [comment]
I have closed issue #179 since I have implemented the functionality to build for both Firefox and Chrome in our dev environment. As a side note, it looks like as a result of this and using NPM and Webpack, I believe we should be able to stop using Grunt to move our libraries into the project repo when there are updates. Since we are using Node packages right from NPM regardless, `npm install` should handle all updates automatically. This relates to issue #182.

**Comment by SebastianZimmeck** [comment]
> npm install should handle all updates

Nice!

**Comment by kalicki1** [comment]
@kbeliauski, @stanleymarkman, I just pushed two different changes that fixes our `storage.js` file and changes its exports. From now on, let's use `storage.get()`, `storage.set()`, etc., instead of `getFromStorage()` and the other old functions. This way we can simplify our imports to `storage` from `storage.js`. 

Pull the changes as you all can.

**Comment by kalicki1** [comment]
As of right now, it looks like we have a few things left to do as part of our refactor. In order of priority, more or less, we have to do the following: 

- [ ] Fix the popup frontend (issue #187)
- [ ] Fix the options page frontend (issue #186)
- [ ] Fix the DOM property functionality (issue #174)
- [ ] Fix dark mode (issue #183) (low priority)

Then in a separate release after the refactor is done: 

- [ ] Upgrade to manifest V3 (issue #175)

Feel free to comment if there's something I missed.

**Comment by kalicki1** [comment]
There are a few outstanding issues related to our refactor, but there are small enough we can spin them off from here and work on them in other issues or in main. Since I merged branch issue-166 (issue #166) into main, I will go ahead and close this issue as well.
